inflate.c

```diff
@@ -120,21 +120,98 @@ int32_t Z_EXPORT PREFIX(inflateReset2)(PREFIX3(stream) *strm, int32_t windowBits
 #endif
     }
 
-    /* set number of window bits, free window if different */
+    /* set number of window bits */
     if (windowBits && (windowBits < MIN_WBITS || windowBits > MAX_WBITS))
         return Z_STREAM_ERROR;
-    if (state->window != NULL && state->wbits != (unsigned)windowBits) {
-        ZFREE_WINDOW(strm, state->window);
-        state->window = NULL;
-    }
 
     /* update state and reset the rest of it */
     state->wrap = wrap;
     state->wbits = (unsigned)windowBits;
     return PREFIX(inflateReset)(strm);
 }
 
-/* This function is hidden in ZLIB_COMPAT builds. */
+#ifdef INF_ALLOC_DEBUG
+#  include <stdio.h>
+#  define LOGSZ(name,size)           fprintf(stderr, "%s is %d bytes\n", name, size)
+#  define LOGSZP(name,size,loc,pad)  fprintf(stderr, "%s is %d bytes, offset %d, padded %d\n", name, size, loc, pad)
+#  define LOGSZPL(name,size,loc,pad) fprintf(stderr, "%s is %d bytes, offset %ld, padded %d\n", name, size, loc, pad)
+#else
+#  define LOGSZ(name,size)
+#  define LOGSZP(name,size,loc,pad)
+#  define LOGSZPL(name,size,loc,pad)
+#endif
+
+/* ===========================================================================
+ * Allocate a big buffer and divide it up into the various buffers inflate needs.
+ * Handles alignment of allocated buffer and alignment of individual buffers.
+ */
+Z_INTERNAL inflate_allocs* alloc_inflate(PREFIX3(stream) *strm) {
+    int currsize = 0;
+
+    /* Define sizes */
+    int window_size = (1 << MAX_WBITS) + 64; /* 64B padding for chunksize */
+    int state_size = sizeof(inflate_state);
+    int alloc_size = sizeof(inflate_allocs);
+
+    /* Calculate relative buffer positions and paddings */
+    LOGSZP("window", window_size, PAD_WINDOW(currsize), PADSZ(currsize,WINDOW_PAD_SIZE));
+    int window_pos = PAD_WINDOW(currsize);
+    currsize = window_pos + window_size;
+
+    LOGSZP("state", state_size, PAD_64(currsize), PADSZ(currsize,64));
+    int state_pos = PAD_64(currsize);
+    currsize = state_pos + state_size;
+
+    LOGSZP("alloc", alloc_size, PAD_16(currsize), PADSZ(currsize,16));
+    int alloc_pos = PAD_16(currsize);
+    currsize = alloc_pos + alloc_size;
+
+    /* Add 64-1 or 4096-1 to allow window alignment, and round size of buffer up to multiple of 64 */
+    int totalsize = PAD_64(currsize + (WINDOW_PAD_SIZE - 1));
```

Comment:
I am not sure about this one.
`totalsize` is related to the "counter" `currsize`.
`*_size` are sizes for individual buffers.

It might be a silly way to differentiate between them, since it is more usual to choose one style or the other.
Alternatively I could rename them to `tsize` and `csize`, but I feel like that loses some of the context.

I can change it to `total_size` and `curr_size` if that is more readable. Thoughts?

Reply:
`total_size` and `curr_size` is more readable imo. I always think that `_` underscore should separate words. It makes it easy say if you were in the future to try to convert from when casing style to another. (hypothetical scenario).

Reply:
Changed